src/DotLiquid/Util/ObjectExtensionMethods.cs

```diff
@@ -52,8 +54,10 @@ public static bool BackCompatSafeTypeInsensitiveEqual(this object value, object
         {
             // NOTE(daviburg): Historically testing for equality cross integer and string boundaries resulted in not equal.
             // This ensures we preserve the behavior.
-            if (value is string && !(otherValue is char) || otherValue is string && !(value is char))
-            {
+            var aType = value.GetType();
+            var bType = otherValue.GetType();
+            if (aType != bType && (_numericTypes.Contains(aType) ^ _numericTypes.Contains(bType)))

```

Comment:
If we do that let's stick to logical operators with || rather than ^

Reply:
This is shorter. Either both are numeric or both are non-numeric. XOR logical operator. https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/boolean-logical-operators